### PR TITLE
Compile NVDA with the Windows 10 SDK

### PR DESCRIPTION
**Pull request by michaelDCurran**
### Summary of the issue:
Currently NVDA is specifically compiled with Windows SDK 7.1A which was the last windows SDK version to support XP. Newer versions of Visual Studio have contained a special re-packaged Windows SDK 7.1a for this purpose. However, this SDK may lack particular features (E.g. static analysis) or optimizations, and will not take advantage of APIs only available on newer Operating Systems. 
 
### Description of how this pull request fixes the issue:
This PR upgrades to SCons 3.0, which now detects and uses Visual Studio 2017.  Our sconscripts specifically check for VS 2017 and will no longer run on anything lower.
This PR also removes the use of Windows SDK 7.1a, thereby falling back to the latest SDK available on the system; This will be the latest Windows 10 SDK.
This PR also  changes the _WIN32_WINNT define to be WIN7 (rather than XP), which allows the use of features available on at most Windows 7.  The linker now also no longer receives /subsystem or /machine arguments, which  now means binaries will default to being suitable for Windows 7 at a minimum, and CPU features such as SSE2 will be used.
There were some calls to GetVersion which is now deprecated. These were either removed (as they were only checking for Vista anyway), or changed to the now recommended versionHelper macros.

### Testing performed:
* NVDA executed on a Windows 7 SP1 machine, tested speak typed characters, read a page in Internet Explorer.

### Known issues with pull request:
None

### Change log entry:
No user visible change.
Changes for Developers:
* NVDA is now compiled with Visual Studio 2017 and the Windows 10 SDK.

### CODE REVIEW

**Comment by josephsl** [comment]
Hi,

When I try to compile this from source, Command prompt and powerShell says MIDL.EXE is not a recognized command or an operable program. Thanks.

**Comment by michaelDCurran** [comment]
It compiled fine on my machine, and on appveyor.

It will require Visual Studio 2015 community as usual. Also In the VS 
2015 community installer, check that you have The latest Windows 10 SDK 
and tools. Specifically:
- Programming Languages:
-- Visual C++
-- * Common Tools for Visual C++ 2015
  - Windows and web development:
-- Universal Windows App Development Tools:
-- * Tools (1.4.1) and Windows 10 SDK (10.0.14393)

You may also want to clean your git working directory as well.

Once Scons 3.0 is released we'll move to Visual Studio 2017 as well.

**Comment by josephsl** [comment]
Yep.

Note to others: do NOT (ever) uninstall whatever SDK that comes with Visual studio Community, otherwise build failures may result.

Thanks.

**Review by feerrenrut — APPROVED** [review]

**Comment by michaelDCurran** [comment]
@feerrenrut: another review please. This now also upgrades to SCons 3.0 official and switches to Visual Studio 2017 for compilation. This completes the switch to more modern tools.

**Comment by leonardder** [comment]
@michaeldcurran: Could you please review the sconscript for nvdahelper/localWin10 There is now some obsolete stuff in there which can be removed.

**Comment by leonardder** [comment]
I also have the following error
```
link /nologo /incremental:no /WX/subsystem:windows,6.01 /release /OPT:REF /export:DllGetClassObject,private /export:DllCanUnloadNow,private /export:GetProxyDllInfo,private /manifest:embed /manifestinput:build\x86_64\ISimpleDOM.manifest /dll /out:build\x86_64\ISimpleDOM.dll /implib:build\x86_64\ISimpleDOM.lib rpcrt4.lib oleaut32.lib ole32.lib /PDB:build\x86_64\ISimpleDOM.dll.pdb /DEBUG build\x86_64\iSimpleDOMNode_i.obj build\x86_64\iSimpleDOMNode_p.obj build\x86_64\iSimpleDOMNode_data.obj
LINK : warning LNK4044: unrecognized option '/WX/subsystem:windows,6.01'; ignored
   Creating library build\x86_64\ISimpleDOM.lib and object build\x86_64\ISimpleDOM.exp
iSimpleDOMNode_data.obj : error LNK2001: unresolved external symbol iSimpleDOMNode_ProxyFileInfo
build\x86_64\ISimpleDOM.dll : fatal error LNK1120: 1 unresolved externals
scons: building terminated because of errors.
```

**Comment by josephsl** [comment]
Hi, also, this means I can safely merge #7520 into this PR to be done by Mick, so no need for me to even submit my own branch. Thanks.

**Comment by michaelDCurran** [comment]
I've now reordered things so that the localWin10 sconscript no longer 
needs to copy much of archBuild_sconscript.

**Comment by michaelDCurran** [comment]
Found and fixed this error as well.
I'm surprised this used to work pre Win10 sdk.

**Review by feerrenrut — APPROVED** [review]

**Comment by leonardder** [comment]
It looks like SCons now has officially moved to github. Would it be possible to use the official repo as a submodule?
https://github.com/SConsProject/scons

**Comment by michaelDCurran** [comment]
Err... this didn't exist when I looked a few weeks ago. Must have just 
happened?

But certainly, if it is official, then yes I'll switch it.

**Comment by leonardder** [comment]
[**@michaelDCurran**](https://github.com/michaelDCurran) commented on [4 okt. 2017 09:37 CEST](https://github.com/nvaccess/nvda/pull/7568#issuecomment-334073678 "2017-10-04T07:37:19Z - Replied by Github Reply Comments"):
> Err... this didn't exist when I looked a few weeks ago. Must have just 
> happened?

1. Probably, yes.

> But certainly, if it is official, then yes I'll switch it.

It is now mentioned on the website as their primary development platform.

**Comment by michaelDCurran** [comment]
Actually, this was too early. Give it another week. master now reset to before this pr. I read the date wrong ;)

**Comment by leonardder** [comment]
Hmm, Github still thinks this is merged.

**Comment by michaelDCurran** [comment]
Yep. Doesn't seem there is any way to fix that.
Can't reopen the pr either.
Even though I reset master to the commit before.

File: nvdaHelper/localWin10/sconscript [hunk 1]
```diff
@@ -65,14 +37,13 @@ localWin10Lib = env.SharedLibrary(
...
 )
-if signExec:
-	env.AddPostAction(localWin10Lib[0], [signExec])
-env.Install(libInstallDir, localWin10Lib)
 
 # UWP dlls can only be dynamically linked with the CRT,
 # but some systems might not have this version of the CRT.
 # Therefore, we must include it.
-vcRedist = os.path.join(progFilesX86, r"Microsoft Visual Studio 14.0\VC\redist\onecore\x86\Microsoft.VC140.CRT")
+vcRedist = os.path.join(progFilesX86, r"Microsoft Visual Studio\2017\Community\VC\Redist\MSVC\14.11.25415\onecore\x86\Microsoft.VC141.CRT")
```
Comment (leonardder) [inline nvdaHelper/localWin10/sconscript]: Would it be possible to get rid of the 14.11.25415 part here by using glob and only use the most recent version available? One problem I see with the current approach is that Appveyor silently updates its Visual Studio 2017 image to a newer build of VS, which might bundle a newer version of the redistributable. Strange enough, when I updated to Visual Studio 2017.4, i lost 14.11.25415 and have now 14.11.25325.

**Comment by michaelDCurran** [comment]
I was able to reproduce this once upgrading VS 2017. I fixed this before 
it was merged to master.

**Pull request merged**